tests/agent_features/test_log_events.py

```diff
@@ -219,3 +250,86 @@ def test():
         record_log_event("A", attributes={"key": "value"})
 
     test()
+
+
+_test_record_log_event_attribute_filtering_params = [
+    ("", "", "context.A", True),
+    ("", "A", "context.A", False),
+    ("", "A", "context.B", True),
+    ("A B", "*", "context.A", True),
+    ("A B", "*", "context.B", True),
+    ("A B", "*", "context.C", False),
+    ("A B", "C", "context.A", True),
+    ("A B", "C", "context.C", False),
+    ("A B", "B", "context.A", True),
+    ("A B", "B", "context.B", False),
+    ("A", "A *", "context.A", False),
+    ("A", "A *", "context.B", False),
+    ("A*", "", "context.A", True),
+    ("A*", "", "context.AB", True),
+    ("", "A*", "context.A", False),
+    ("", "A*", "context.B", True),
+    ("A*", "AB", "context.AC", True),
+    ("A*", "AB", "context.AB", False),
+    ("AB", "A*", "context.AB", True),
+    ("A*", "AB*", "context.ACB", True),
+    ("A*", "AB*", "context.ABC", False),
+    # Linking attributes not affected by filtering
+    ("", "", "entity.name", True),
+    ("A", "*", "entity.name", True),
+    ("", "*", "entity.name", True),
+]
+
+
+@pytest.mark.parametrize("include,exclude,attr,expected", _test_record_log_event_attribute_filtering_params)
+def test_record_log_event_attribute_filtering_inside_transaction(include, exclude, attr, expected):
+    if expected:
+        expected_event = {"required_attrs": [attr]}
+    else:
+        expected_event = {"forgone_attrs": [attr]}
+
+    @override_application_settings(
+        {
+            "application_logging.forwarding.enabled": True,
+            "application_logging.forwarding.context_data.enabled": True,
+            "application_logging.forwarding.context_data.include": _parse_attributes(include),
+            "application_logging.forwarding.context_data.exclude": _parse_attributes(exclude),
+        }
+    )
+    @validate_log_events(**expected_event)
+    @validate_log_event_count(1)
+    @background_task()
+    def test():
+        if attr.startswith("context."):
+            record_log_event("A", attributes={attr.lstrip("context."): 1})
+        else:
+            record_log_event("A")
+
+    test()
+
+
```

Comment:
Do you think we need to test the non-datatype-specific `attributes.include` and `attributes.exclude` here at all?

Reply:
Probably not, they're pretty well tested in their own section on the attribute filter rules in the cross agent tests. This implementation uses that system so we shouldn't need to reproduce those tests as well.

We could be a lot less thorough here if we wanted and rely on the precedence definitions in the cross agent tests. I could simplify this to just a couple runs, but it's not really hurting anything by staying as these are quite fast.